sql/parser/eval.go

```diff
@@ -528,27 +529,7 @@ var evalTupleIN = cmpOp{
 
 		vtuple := values.(DTuple)
 
-		// TODO(pmattis): If we're evaluating the expression multiple times we should
-		// use a map when possible. This works as long as arg is not a tuple. Note
-		// that the usage of the map is currently disabled via the "&& false" because
-		// building the map is a pessimization if we're only evaluating the
-		// expression once. We need to determine when the expression will be
-		// evaluated multiple times before enabling. Also need to figure out a way to
-		// use the map approach for tuples. One idea is to encode the tuples into
-		// strings and then use a map of strings.
-		if _, ok := arg.(DTuple); !ok && false {
-			m := make(map[Datum]struct{}, len(vtuple))
-			for _, val := range vtuple {
-				if reflect.TypeOf(arg) != reflect.TypeOf(val) {
-					return DummyBool, fmt.Errorf("unsupported comparison operator: <%s> %s <%s>",
-						arg.Type(), EQ, val.Type())
-				}
-				m[val] = struct{}{}
-			}
-			if _, exists := m[arg]; exists {
-				return DBool(true), nil
-			}
-		} else {
+		if _, ok := arg.(DTuple); ok {
```

Comment:
I didn't realize that the order was supposed to be preserved for values in inner tuples. Because it is (eg. `((1,2)) NotIn ((2,1))`), the binary search will work. Removed the block.